Add a `LayoutProps` type so fs-router layouts can type their (now-passed) route params

## Summary

Layouts now receive route params from their own path at runtime, but there is no typed way to access them. `waku/router` exports `PageProps` but no `LayoutProps`, and `PageProps` cannot be reused for a layout because it has no `children`. So fs-router layouts must hand-type their params (`{ aaa: string; children: ReactNode }`), which is not derived from the route and gets no type safety.

The exact prop shape is already computed internally for the `createPages` API (`CreateLayout`). This proposes exposing it as a `LayoutProps<Path>` helper so fs-router layouts get the same typed-params DX that `PageProps` gives pages, including the correct parent-path-only param scope.

## Background: layouts now receive parent-path params at runtime

A layout receives the params of its own path. A nested example makes the scope clear:

```
./src/pages/foo/[aaa]/[bbb].tsx   -> page   at /foo/[aaa]/[bbb], gets { aaa, bbb }
./src/pages/foo/[aaa]/_layout.tsx -> layout at /foo/[aaa],        gets { aaa } only
```

The page is deeper than the layout, so it has both slugs; the layout sits at `/foo/[aaa]`, so it only has `aaa`. It cannot see the child param `bbb`.

Verified at runtime by rendering the prop keys each component receives. Request `/foo/x/y`:

```
status:      200
layout keys: "aaa"            (layout aaa = "x")
page keys:   "aaa,bbb,path"   (page aaa = "x", bbb = "y")
```

So the layout receives `aaa` only, while the page receives `aaa` and `bbb`.

## The type gap

There is no `LayoutProps` to type those params, and `PageProps` does not fit because it has no `children`:

```tsx
// src/pages/foo/[aaa]/_layout.tsx
import type { PageProps } from 'waku/router';
export default function FooLayout({ aaa, children }: PageProps<'/foo/[aaa]'>) { /* ... */ }
```
```
error TS2339: Property 'children' does not exist on type
  '{ path: `/foo/${string}`; query: string; aaa: string; }'.
```

So the only option today is to hand-type the props:

```tsx
import type { ReactNode } from 'react';
export default function FooLayout({ aaa, children }: { aaa: string; children: ReactNode }) { /* ... */ }
```

This type-checks, but `aaa` is hand-written rather than derived from the route, so there is no validation that the layout's params match its path, and the annotation drifts if the route changes.

## The shape already exists internally

`CreateLayout` already types the layout component with the right shape:

```ts
// waku/dist/router/create-pages.d.ts (CreateLayout)
component: FunctionComponent<
  { children: ReactNode } & Omit<PropsForPages<Path>, 'path' | 'query'>
>;
```

That is exactly what a `LayoutProps<Path>` helper needs; it just is not exposed for fs-router use.

## Proposal

Export a `LayoutProps` from `waku/router`, mirroring the `CreateLayout` component prop shape:

```ts
import type { ReactNode } from 'react';

export type LayoutProps<Path extends PagePath<CreatePagesConfig> | (string & {})> =
  { children: ReactNode } & Omit<PropsForPages<Path>, 'path' | 'query'>;
```

A fs-router layout is then typed like a page, with `Path` set to the layout's own path:

```tsx
// src/pages/foo/[aaa]/_layout.tsx
import type { LayoutProps } from 'waku/router';
export default function FooLayout({ aaa, children }: LayoutProps<'/foo/[aaa]'>) {
  return <div>{aaa}{children}</div>;
}
```

Because `LayoutProps` is keyed by the layout's own path, the parent-path-only scope is captured automatically: slug extraction over `'/foo/[aaa]'` only sees `aaa`, never `bbb`. No special handling is required.

## Experimental results

Prototyped by adding the `LayoutProps` export above to the installed `base-types.d.ts` and type-checking the real app (`tsc --noEmit`).

Resolved type (note: only the parent param, plus children):

```
LayoutProps<'/foo/[aaa]'>          =>  { children: ReactNode; aaa: string }
PageProps<'/foo/[aaa]/[bbb]'>      =>  { path; query; aaa: string; bbb: string }
```

The parent-path-only scope holds in the types, matching the runtime. Accessing the child param from the layout type errors:

```tsx
const layout = (p: LayoutProps<'/foo/[aaa]'>) => { p.aaa; p.children; p.bbb; };
//                                                                     ^^^^^
// error TS2339: Property 'bbb' does not exist on type 'LayoutProps<"/foo/[aaa]">'.

const page = (p: PageProps<'/foo/[aaa]/[bbb]'>) => { p.aaa; p.bbb; }; // both ok
```

Other checks:

- Typing a layout with `LayoutProps<'/foo/[aaa]'>`: clean; `aaa` derived from the route, `children` typed.
- Wrong param name errors:
  ```
  error TS2339: Property 'nope' does not exist on type 'LayoutProps<"/foo/[aaa]">'.
  ```
- `path` and `query` are omitted from the layout surface:
  ```
  error TS2339: Property 'path' does not exist on type 'LayoutProps<"/foo/[aaa]">'.
  error TS2339: Property 'query' does not exist on type 'LayoutProps<"/foo/[aaa]">'.
  ```

## Notes / open questions

- **Path universe for the constraint.** `PagePath<CreatePagesConfig>` lists *page* paths. A layout can exist at a path with no co-located page (e.g. `src/pages/foo/_layout.tsx` wrapping pages only under `/foo/[aaa]`). Such a layout-only path would not be in the page-path union, so the constraint and the generated `pages.gen.ts` may need to surface layout paths as well for full route validation. The prototype above worked because the test layout shared a path with a page; the `(string & {})` arm accepts anything regardless.
- **Route-literal validation.** `LayoutProps` should get the same constraint treatment proposed for `PageProps` (only fall back to `string` when there are no known routes), so `LayoutProps<'/foo/[wrong]'>` is validated once routes are known. See the related `PageProps` route-validation issue.
- **`ReactNode` import.** `base-types.d.ts` would need `import type { ReactNode } from 'react'` (or an inline `import('react').ReactNode`). `react` is already a peer dependency.

## Environment

- `waku@1.0.0-beta.3` (please confirm against latest)
- TypeScript `6.0.x`, `--strict`
- Verified with `waku dev` (runtime) and `tsc --noEmit` (types)

--- a/e2e/fixtures/fs-router/src/layout-props-typing.ts
+++ b/e2e/fixtures/fs-router/src/layout-props-typing.ts
@@ -1,0 +1,18 @@
+import type { LayoutProps } from 'waku/router';
+
+// Type-level assertions only; never called. CreatePagesConfig.layouts is
+// augmented by the generated pages.gen.ts, so a known layout path (including a
+// layout-only path with no co-located page, e.g. /cache-check) resolves while
+// an unknown path, or a page path that has no layout, is rejected.
+export function assertLayoutPropsTyping(
+  rootProps: LayoutProps<'/'>,
+  layoutOnlyProps: LayoutProps<'/cache-check'>,
+) {
+  void rootProps.children;
+  void layoutOnlyProps.children;
+}
+
+// @ts-expect-error an unknown layout path is not a known layout
+export type InvalidLayoutProps = LayoutProps<'/no-such-layout'>;
+// @ts-expect-error a page path with no co-located layout is not a layout path
+export type PageOnlyLayoutProps = LayoutProps<'/nested/[name]'>;

--- a/e2e/fixtures/fs-router/src/layout-props-typing.ts
+++ b/e2e/fixtures/fs-router/src/layout-props-typing.ts
@@ -2,14 +2,22 @@ import type { LayoutProps } from 'waku/router';
 
 // Type-level assertions only; never called. CreatePagesConfig.layouts is
 // augmented by the generated pages.gen.ts, so a known layout path (including a
-// layout-only path with no co-located page, e.g. /cache-check) resolves while
-// an unknown path, or a page path that has no layout, is rejected.
+// layout-only path with no co-located page, e.g. /cache-check or the dynamic
+// /dynamic-layout/[id]) resolves while an unknown path, or a page path that has
+// no layout, is rejected.
 export function assertLayoutPropsTyping(
   rootProps: LayoutProps<'/'>,
   layoutOnlyProps: LayoutProps<'/cache-check'>,
+  dynamicProps: LayoutProps<'/dynamic-layout/[id]'>,
 ) {
   void rootProps.children;
   void layoutOnlyProps.children;
+  void dynamicProps.children;
+  // a dynamic layout path derives its own param via RouteParams
+  const id: string = dynamicProps.id;
+  void id;
+  // @ts-expect-error a child page param ([sub]) does not leak into the layout
+  void dynamicProps.sub;
 }
 
 // @ts-expect-error an unknown layout path is not a known layout

--- a/e2e/fixtures/fs-router/src/pages/cache-check/_layout.tsx
+++ b/e2e/fixtures/fs-router/src/pages/cache-check/_layout.tsx
@@ -1,6 +1,6 @@
-import type { ReactNode } from 'react';
+import type { LayoutProps } from 'waku/router';
 
-const CacheCheckLayout = ({ children }: { children: ReactNode }) => {
+const CacheCheckLayout = ({ children }: LayoutProps<'/cache-check'>) => {
   // Render Date.now() so an e2e test can detect whether the (static)
   // layout was rendered at build time or live at runtime by comparing
   // against a timestamp captured before the request.

--- a/e2e/fixtures/fs-router/src/pages/dynamic-layout/[id]/[sub].tsx
+++ b/e2e/fixtures/fs-router/src/pages/dynamic-layout/[id]/[sub].tsx
@@ -1,0 +1,17 @@
+import type { PageProps } from 'waku/router';
+
+const Page = ({ id, sub }: PageProps<'/dynamic-layout/[id]/[sub]'>) => (
+  <div>
+    <h2>
+      {id} / {sub}
+    </h2>
+  </div>
+);
+
+export const getConfig = () => {
+  return {
+    render: 'dynamic',
+  } as const;
+};
+
+export default Page;

--- a/e2e/fixtures/fs-router/src/pages/dynamic-layout/[id]/_layout.tsx
+++ b/e2e/fixtures/fs-router/src/pages/dynamic-layout/[id]/_layout.tsx
@@ -1,0 +1,13 @@
+import type { LayoutProps } from 'waku/router';
+
+const DynamicLayout = ({
+  id,
+  children,
+}: LayoutProps<'/dynamic-layout/[id]'>) => (
+  <div>
+    <span data-testid="dynamic-layout-id">{id}</span>
+    {children}
+  </div>
+);
+
+export default DynamicLayout;

--- a/packages/waku/src/lib/vite-plugins/fs-router-typegen.ts
+++ b/packages/waku/src/lib/vite-plugins/fs-router-typegen.ts
@@ -231,6 +231,7 @@ export async function generateFsRouterTypes(pagesDir: string) {
 
   const generateFile = async (filePaths: string[]): Promise<string | null> => {
     const fileInfo: { path: string; src: string; hasGetConfig: boolean }[] = [];
+    const layoutPaths: string[] = [];
     const moduleNames = getImportModuleNames(filePaths);
 
     for (const filePath of filePaths) {
@@ -243,11 +244,11 @@ export async function generateFsRouterTypes(pagesDir: string) {
         return null;
       }
 
-      if (
-        filePath.endsWith('/_layout.tsx') ||
-        isIgnoredPath(filePath.split('/'))
-      ) {
+      if (isIgnoredPath(filePath.split('/'))) {
         continue;
+      } else if (filePath.endsWith('/_layout.tsx')) {
+        const path = filePath.slice(0, -'/_layout.tsx'.length);
+        layoutPaths.push(getGrouplessPath(path) || '/');
       } else if (filePath.endsWith('/index.tsx')) {
         const path = filePath.slice(0, -'/index.tsx'.length);
         fileInfo.push({
@@ -301,6 +302,15 @@ export async function generateFsRouterTypes(pagesDir: string) {
     }
     lines[lines.length - 1] += ';';
 
+    const uniqueLayoutPaths = [...new Set(layoutPaths)];
+    if (uniqueLayoutPaths.length) {
+      lines.push('', '// prettier-ignore', 'type Layout =');
+      for (const path of uniqueLayoutPaths) {
+        lines.push(`| { path: '${path}' }`);
+      }
+      lines[lines.length - 1] += ';';
+    }
+
     lines.push(
       '',
       '// prettier-ignore',
@@ -310,6 +320,7 @@ export async function generateFsRouterTypes(pagesDir: string) {
       '  }',
       '  interface CreatePagesConfig {',
       '    pages: Page;',
+      ...(uniqueLayoutPaths.length ? ['    layouts: Layout;'] : []),
       '  }',
       ...(hasAnyGetConfig
         ? [

--- a/packages/waku/src/router/base-types.ts
+++ b/packages/waku/src/router/base-types.ts
@@ -1,4 +1,6 @@
+import type { ReactNode } from 'react';
 import type {
+  LayoutPath,
   PagePath,
   PropsForPages,
 } from './create-pages-utils/inferred-path-types.js';
@@ -31,3 +33,14 @@ export type PageProps<
     ? string
     : PagePath<CreatePagesConfig>,
 > = PropsForPages<Path>;
+
+/**
+ * Props for layouts when using `createPages`. Adds the required `children` and
+ * drops `path`/`query`. Validated against the generated layout paths, which can
+ * include layout-only paths that have no co-located page.
+ */
+export type LayoutProps<
+  Path extends [LayoutPath<CreatePagesConfig>] extends [never]
+    ? string
+    : LayoutPath<CreatePagesConfig>,
+> = { children: ReactNode } & Omit<PropsForPages<Path>, 'path' | 'query'>;

--- a/packages/waku/src/router/base-types.ts
+++ b/packages/waku/src/router/base-types.ts
@@ -3,6 +3,7 @@ import type {
   LayoutPath,
   PagePath,
   PropsForPages,
+  RouteParams,
 } from './create-pages-utils/inferred-path-types.js';
 import type { Prettify } from './create-pages-utils/util-types.js';
 
@@ -36,18 +37,13 @@ export type PageProps<
 > = PropsForPages<Path>;
 
 /**
- * Props for layouts when using `createPages`. Adds the required `children` and
- * keeps only the layout's own route params: a layout receives neither `path`,
- * `query`, nor `search` at runtime. Validated against the generated layout
- * paths, which can include layout-only paths that have no co-located page.
+ * Props for layouts when using `createPages`. Adds the required `children` to
+ * the layout's own route params (a layout receives neither `path`, `query`, nor
+ * `search` at runtime). Validated against the generated layout paths, which can
+ * include layout-only paths that have no co-located page.
  */
 export type LayoutProps<
   Path extends [LayoutPath<CreatePagesConfig>] extends [never]
     ? string
     : LayoutPath<CreatePagesConfig>,
-> = Prettify<
-  { children: ReactNode } & Omit<
-    PropsForPages<Path>,
-    'path' | 'query' | 'search'
-  >
->;
+> = Prettify<{ children: ReactNode } & RouteParams<Path>>;

--- a/packages/waku/src/router/base-types.ts
+++ b/packages/waku/src/router/base-types.ts
@@ -4,6 +4,7 @@ import type {
   PagePath,
   PropsForPages,
 } from './create-pages-utils/inferred-path-types.js';
+import type { Prettify } from './create-pages-utils/util-types.js';
 
 // PathsForPages collects the paths from the `createPages` result type
 // GetConfigResponse is a helper for generating types with fs-router
@@ -36,11 +37,17 @@ export type PageProps<
 
 /**
  * Props for layouts when using `createPages`. Adds the required `children` and
- * drops `path`/`query`. Validated against the generated layout paths, which can
- * include layout-only paths that have no co-located page.
+ * keeps only the layout's own route params: a layout receives neither `path`,
+ * `query`, nor `search` at runtime. Validated against the generated layout
+ * paths, which can include layout-only paths that have no co-located page.
  */
 export type LayoutProps<
   Path extends [LayoutPath<CreatePagesConfig>] extends [never]
     ? string
     : LayoutPath<CreatePagesConfig>,
-> = { children: ReactNode } & Omit<PropsForPages<Path>, 'path' | 'query'>;
+> = Prettify<
+  { children: ReactNode } & Omit<
+    PropsForPages<Path>,
+    'path' | 'query' | 'search'
+  >
+>;

--- a/packages/waku/src/router/create-pages-utils/inferred-path-types.ts
+++ b/packages/waku/src/router/create-pages-utils/inferred-path-types.ts
@@ -185,6 +185,13 @@ export type PagePath<Config> = Config extends {
   ? Path
   : never;
 
+/** Layout paths with slugs as string literals */
+export type LayoutPath<Config> = Config extends {
+  layouts: { path: infer Path };
+}
+  ? Path
+  : never;
+
 type IndividualSlugType<Slug extends string> = Slug extends `...${string}`
   ? string[]
   : string;

--- a/packages/waku/src/router/create-pages.tsx
+++ b/packages/waku/src/router/create-pages.tsx
@@ -266,7 +266,10 @@ export type CreateLayout = <Path extends string>(
         render: 'dynamic';
         path: Path;
         component: FunctionComponent<
-          { children: ReactNode } & Omit<PropsForPages<Path>, 'path' | 'query'>
+          { children: ReactNode } & Omit<
+            PropsForPages<Path>,
+            'path' | 'query' | 'search'
+          >
         >;
         unstable_sourceFile?: string | undefined;
       }
@@ -274,7 +277,10 @@ export type CreateLayout = <Path extends string>(
         render: 'static';
         path: Path;
         component: FunctionComponent<
-          { children: ReactNode } & Omit<PropsForPages<Path>, 'path' | 'query'>
+          { children: ReactNode } & Omit<
+            PropsForPages<Path>,
+            'path' | 'query' | 'search'
+          >
         >;
         unstable_sourceFile?: string | undefined;
       }
@@ -286,7 +292,9 @@ export type CreateLayout = <Path extends string>(
      * Per-render element tag (etag) for a `dynamic` layout. Ignored for `static`
      * layouts.
      */
-    unstable_getEtag?: GetEtag<Omit<PropsForPages<Path>, 'path' | 'query'>>;
+    unstable_getEtag?: GetEtag<
+      Omit<PropsForPages<Path>, 'path' | 'query' | 'search'>
+    >;
   },
 ) => void;
 

--- a/packages/waku/src/router/create-pages.tsx
+++ b/packages/waku/src/router/create-pages.tsx
@@ -17,6 +17,7 @@ import type {
   AnyPage,
   GetSlugs,
   PropsForPages,
+  RouteParams,
   Unstable_SearchCodec,
 } from './create-pages-utils/inferred-path-types.js';
 import { unstable_defineRouter } from './define-router.js';
@@ -266,10 +267,7 @@ export type CreateLayout = <Path extends string>(
         render: 'dynamic';
         path: Path;
         component: FunctionComponent<
-          { children: ReactNode } & Omit<
-            PropsForPages<Path>,
-            'path' | 'query' | 'search'
-          >
+          { children: ReactNode } & RouteParams<Path>
         >;
         unstable_sourceFile?: string | undefined;
       }
@@ -277,10 +275,7 @@ export type CreateLayout = <Path extends string>(
         render: 'static';
         path: Path;
         component: FunctionComponent<
-          { children: ReactNode } & Omit<
-            PropsForPages<Path>,
-            'path' | 'query' | 'search'
-          >
+          { children: ReactNode } & RouteParams<Path>
         >;
         unstable_sourceFile?: string | undefined;
       }
@@ -292,9 +287,7 @@ export type CreateLayout = <Path extends string>(
      * Per-render element tag (etag) for a `dynamic` layout. Ignored for `static`
      * layouts.
      */
-    unstable_getEtag?: GetEtag<
-      Omit<PropsForPages<Path>, 'path' | 'query' | 'search'>
-    >;
+    unstable_getEtag?: GetEtag<RouteParams<Path>>;
   },
 ) => void;
 

--- a/packages/waku/tests/layout-props-types.test.ts
+++ b/packages/waku/tests/layout-props-types.test.ts
@@ -19,6 +19,11 @@ describe('LayoutProps', () => {
     expectType<TypeEqual<'query' extends keyof Props ? true : false, false>>(
       true,
     );
+    // parent-path-only scope: a child page param (bbb of /foo/[aaa]/[bbb])
+    // never leaks into the layout's props
+    expectType<TypeEqual<'bbb' extends keyof Props ? true : false, false>>(
+      true,
+    );
     const props: Props = { children: null, aaa: 'x' };
     expect(props.aaa).toBe('x');
   });

--- a/packages/waku/tests/layout-props-types.test.ts
+++ b/packages/waku/tests/layout-props-types.test.ts
@@ -1,0 +1,25 @@
+import type { ReactNode } from 'react';
+import { expectType } from 'ts-expect';
+import type { TypeEqual } from 'ts-expect';
+import { describe, expect, it } from 'vitest';
+import type { LayoutProps } from '../src/router/base-types.js';
+
+// In this suite CreatePagesConfig is not augmented, so LayoutProps accepts any
+// string path. Layouts use a loose constraint on purpose: a layout path may
+// have no co-located page, so it is not necessarily in the page route union
+// (e.g. a layout at /foo/[aaa] whose only page is /foo/[aaa]/[bbb]).
+describe('LayoutProps', () => {
+  it('adds children, derives its own params, and drops path/query', () => {
+    type Props = LayoutProps<'/foo/[aaa]'>;
+    expectType<TypeEqual<Props['children'], ReactNode>>(true);
+    expectType<TypeEqual<Props['aaa'], string>>(true);
+    expectType<TypeEqual<'path' extends keyof Props ? true : false, false>>(
+      true,
+    );
+    expectType<TypeEqual<'query' extends keyof Props ? true : false, false>>(
+      true,
+    );
+    const props: Props = { children: null, aaa: 'x' };
+    expect(props.aaa).toBe('x');
+  });
+});

--- a/packages/waku/tests/layout-props-types.test.ts
+++ b/packages/waku/tests/layout-props-types.test.ts
@@ -2,14 +2,26 @@ import type { ReactNode } from 'react';
 import { expectType } from 'ts-expect';
 import type { TypeEqual } from 'ts-expect';
 import { describe, expect, it } from 'vitest';
-import type { LayoutProps } from '../src/router/base-types.js';
+import type {
+  LayoutProps,
+  Unstable_SearchCodec,
+} from '../src/router/base-types.js';
+import type { PropsForPages } from '../src/router/create-pages-utils/inferred-path-types.js';
 
-// In this suite CreatePagesConfig is not augmented, so LayoutProps accepts any
-// string path. Layouts use a loose constraint on purpose: a layout path may
-// have no co-located page, so it is not necessarily in the page route union
-// (e.g. a layout at /foo/[aaa] whose only page is /foo/[aaa]/[bbb]).
+// In this suite CreatePagesConfig is not augmented, so LayoutPath resolves to
+// never and LayoutProps falls back to accepting any string (the pre-codegen
+// behavior, like PageProps). Validation against generated layout paths is proven
+// in the augmented fs-router fixture (layout-props-typing.ts). A search codec is
+// registered below to prove the layout drops `search` even when a page at the
+// same path would receive it.
+declare module '../src/router/base-types.js' {
+  interface SearchCodecsConfig {
+    '/foo/[aaa]': Unstable_SearchCodec<{ q: string }>;
+  }
+}
+
 describe('LayoutProps', () => {
-  it('adds children, derives its own params, and drops path/query', () => {
+  it('adds children, derives its own params, and drops path/query/search', () => {
     type Props = LayoutProps<'/foo/[aaa]'>;
     expectType<TypeEqual<Props['children'], ReactNode>>(true);
     expectType<TypeEqual<Props['aaa'], string>>(true);
@@ -22,6 +34,13 @@ describe('LayoutProps', () => {
     // parent-path-only scope: a child page param (bbb of /foo/[aaa]/[bbb])
     // never leaks into the layout's props
     expectType<TypeEqual<'bbb' extends keyof Props ? true : false, false>>(
+      true,
+    );
+    // a page at this path receives a typed `search`, but the layout does not
+    expectType<TypeEqual<PropsForPages<'/foo/[aaa]'>['search'], { q: string }>>(
+      true,
+    );
+    expectType<TypeEqual<'search' extends keyof Props ? true : false, false>>(
       true,
     );
     const props: Props = { children: null, aaa: 'x' };

--- a/packages/waku/tests/vite-plugin-fs-router-typegen.test.ts
+++ b/packages/waku/tests/vite-plugin-fs-router-typegen.test.ts
@@ -84,6 +84,11 @@ import type { getConfig as File_ØnéTwoThree_getConfig } from './pages/øné_tw
     expect(generated).toContain(
       'interface SearchCodecsConfig extends SearchCodecsForPages<Page> {}',
     );
+    // layout paths (including dynamic ones) are emitted into a Layout union
+    expect(generated).toContain('type Layout =');
+    expect(generated).toContain("| { path: '/' }");
+    expect(generated).toContain("| { path: '/[category]' }");
+    expect(generated).toContain('layouts: Layout;');
   });
 
   test('generates types when waku.server uses fsRouter (managed mode)', async () => {
@@ -172,7 +177,7 @@ import type { getConfig as File_ØnéTwoThree_getConfig } from './pages/øné_tw
     expect(generated).toContain("| { path: '/about'; render: 'static' }");
   });
 
-  test('generates paths while skipping ignored/layout files and missing getConfig', async () => {
+  test('emits page and layout paths while skipping ignored files and missing getConfig', async () => {
     const generated = await generateFsRouterTypes(
       path.join(fixturesDir, 'plugin-fs-router-typegen-complex', 'pages'),
     );
@@ -187,6 +192,7 @@ import type { getConfig as File_ØnéTwoThree_getConfig } from './pages/øné_tw
     expect(generated).toContain(
       `import type { getConfig as File_DocsIndex_getConfig } from './pages/docs/index';`,
     );
+    // layout files are not imported and their filename is never leaked
     expect(generated).not.toContain('_layout');
     expect(generated).not.toContain('_components');
     expect(generated).toContain(
@@ -199,6 +205,10 @@ import type { getConfig as File_ØnéTwoThree_getConfig } from './pages/øné_tw
       "| ({ path: '/admin/dashboard' } & GetConfigResponse<typeof File_AdminDashboard_getConfig>)",
     );
     expect(generated).toContain("| { path: '/blog/[slug]'; render: 'static' }");
+    // the layout at docs/_layout.tsx is emitted as a layout path
+    expect(generated).toContain('type Layout =');
+    expect(generated).toContain("| { path: '/docs' }");
+    expect(generated).toContain('layouts: Layout;');
     expect(generated).toContain("declare module 'waku/router'");
   });
 });

